File: examples/snippets/modules/ROOT/examples/akkastreams-scala/src/main/scala/cloudflow/akkastreamsdoc/MultiOutlet.scala
package cloudflow.akkastreamsdoc

// tag::sink2[]
import cloudflow.akkastream._
import cloudflow.akkastream.scaladsl._
import cloudflow.akkastream.util.scaladsl._

import cloudflow.streamlets._
import cloudflow.streamlets.avro._

import scala.collection.immutable

class MultiOutlet extends AkkaStreamlet {
  val in      = AvroInlet[Data]("in")
  val invalid = AvroOutlet[DataInvalid]("invalid").withPartitioner(data ⇒ data.key)
  val valid   = AvroOutlet[Data]("valid").withPartitioner(RoundRobinPartitioner)
  val shape   = StreamletShape(in).withOutlets(invalid, valid)

  override def createLogic = new RunnableGraphStreamletLogic() {
    def runnableGraph =
      sourceWithCommittableContext(in)
        .map { data =>
          if (data.value < 0) (immutable.Seq(DataInvalid(data.key, data.value, "All data must be positive numbers!")), immutable.Seq.empty)
          else (immutable.Seq.empty, immutable.Seq(data))
        }
        .to(MultiOutlet.sink(invalid, valid))
  }
}
// end::sink2[]

File: examples/snippets/modules/ROOT/examples/akkastreams-java/src/main/java/cloudflow/akkastreamsdoc/MultiOutlet.java
package cloudflow.akkastreamsdoc;

import java.util.Collections;
import java.util.List;

import akka.stream.javadsl.*;

import akka.NotUsed;
import akka.actor.*;
import akka.kafka.ConsumerMessage.Committable;
import akka.stream.*;

import com.typesafe.config.Config;

// tag::sink2[]
import cloudflow.streamlets.*;
import cloudflow.streamlets.avro.*;
import cloudflow.akkastream.*;
import cloudflow.akkastream.javadsl.*;
import cloudflow.akkastream.util.javadsl.*;

import akka.japi.Pair;

public class MultiOutlet extends AkkaStreamlet {
  AvroInlet<Data> inlet = AvroInlet.<Data>create("in", Data.class);
  AvroOutlet<DataInvalid> invalidOutlet =
      AvroOutlet.<DataInvalid>create(
          "invalid", d -> d.getKey(), DataInvalid.class);
  AvroOutlet<Data> validOutlet =
      AvroOutlet.<Data>create(
          "valid", d -> RoundRobinPartitioner.apply(d), Data.class);

  public StreamletShape shape() {
    return StreamletShape.createWithInlets(inlet).withOutlets(invalidOutlet, validOutlet);
  }

  public AkkaStreamletLogic createLogic() {
    return new RunnableGraphStreamletLogic(getContext()) {
      public RunnableGraph createRunnableGraph() {
        return getSourceWithCommittableContext(inlet)
                .map(data -> {
                        final List<DataInvalid> invalid;
                        final List<Data> valid;
                        if (data.getValue() < 0) {
                            invalid = Collections.singletonList(new DataInvalid(data.getKey(), data.getValue(), "All data must be positive numbers!"));
                            valid = Collections.emptyList();
                        } else {
                            invalid = Collections.emptyList();
                            valid = Collections.singletonList(data);
                        }
                        return Pair.create(invalid, valid);
                })
                .to(MultiOutlet.sink2(invalidOutlet, validOutlet, getContext()));
      }
    };
  }
}
// end::sink2[]
